x/staking/keeper/validator.go

```diff
@@ -183,31 +179,31 @@ func (k Keeper) RemoveValidator(ctx sdk.Context, address sdk.ValAddress) {
 	if !found {
 		return
 	}
+
 	if validator.Status != sdk.Unbonded {
 		panic("Cannot call RemoveValidator on bonded or unbonding validators")
 	}
 
-	// if any tokens remain, remove from pool (burning the tokens).
-	// this happens if shares are zero but tokens are not.
-	// TODO: Remove once https://github.com/cosmos/cosmos-sdk/pull/2958 is merged
-	pool := k.GetPool(ctx)
-	pool.NotBondedTokens = pool.NotBondedTokens.Sub(validator.Tokens)
-	k.SetPool(ctx, pool)
+	if validator.Tokens.GT(sdk.ZeroInt()) {
+		panic("validator being removed should never have positive tokens")
+	}
 
 	// delete the old validator record
 	store := ctx.KVStore(k.storeKey)
 	store.Delete(GetValidatorKey(address))
 	store.Delete(GetValidatorByConsAddrKey(sdk.ConsAddress(validator.ConsPubKey.Address())))
 	store.Delete(GetValidatorsByPowerIndexKey(validator))
 
+	// delete the queue record
+	k.DeleteValidatorQueue(ctx, validator)
```

Comment:
hmm I prefer (mild) to leave all the queue logic in the queue... I think it muddies the code to have these buried around. thoughts?

Reply:
I have a mild preference for the opposite, it's strange to keep a reference to a nonexistent record and deleting earlier means that storage is not used for as long